### PR TITLE
[Only] Deprecate Persistent mode of Cluster Sharding 

### PR DESCRIPTION
**Pull request by helena**
## References 
#27528 
## Background
Because we want to deprecate persistence mode for Cluster Sharding, but the replacement is non-trivial and we prefer to not allocate the time in order to get 2.6, and the alternative is wait until 2.7 to do this all at once, we decided to push the deprecation in for 2.6 and the replacement in post-2.6.

This includes deprecation and modification of the docs where it makes sense - the rest of the doc changes belong in the replacement PR.

Note: I went back and forth on removing this section https://doc.akka.io/docs/akka/snapshot/typed/cluster-sharding.html#persistence-example to not encourage new users of sharding to choose/learn to use persistence mode. Finally decided it is worth being in until the replacement to support existing users wanting a reference.

There are quite a few doc updates to do in the later PR. I tried to do only what made sense, along with some cleanup and clarification.

### CODE REVIEW

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ClusterShardingSettings.scala [hunk 1]
```diff
@@ -106,6 +106,7 @@ object ClusterShardingSettings {
         throw new IllegalArgumentException(
           "Not recognized StateStoreMode, only 'persistence' and 'ddata' are supported.")
   }
+
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ClusterShardingSettings.scala]: For typed I think this is what you'd use to programmatically set, so makes sense to deprecate, for the untyped I think a user won't use the constant, it's just used internally so it probably doesn't help to deprecate that.

Perhaps log a deprecation warning if sharding is started with the persistent mode enabled?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 2]
```diff
@@ -649,6 +649,29 @@ private[akka] trait RememberingShard {
   }
 }
 
+/** INTERNAL API */
+private[akka] object PersistentShard {
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: `@InternalApi` also

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 3]
```diff
@@ -658,6 +681,7 @@ private[akka] trait RememberingShard {
  *
  * @see [[ClusterSharding$ ClusterSharding extension]]
  */
+@deprecated("Use `ddata` mode, persistence mode is deprecated.", "2.6.0")
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: No need to deprecate the internal API, those we don't promise any compatibility across versions for anyway.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 4]
```diff
@@ -38,6 +39,7 @@ object ShardCoordinator {
    * INTERNAL API
    * Factory method for the [[akka.actor.Props]] of the [[ShardCoordinator]] actor.
    */
+  @silent("deprecated")
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Same as above, no user will ever call this and see the deprecation

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 5]
```diff
@@ -274,16 +276,19 @@ shard resolution, e.g. to avoid too fine grained shards. Once a shard's location
...
-The state of the coordinator and the state of [Remembering Entities](#cluster-sharding-remembering) of the shards
-are persistent (durable) to survive failures. @ref:[Distributed Data](distributed-data.md) or @ref:[Persistence](persistence.md)
-can be used for the storage. Distributed Data is used by default.
+There are two cluster sharding states managed:
+1. @ref:[ShardCoordinator State](#shardcoordinator-state) - the `Shard` locations
+1. @ref:[Remembering Entities](#remembering-entities) - the entities in each `Shard`, which is optional, and disabled by default
+ 
+For these, there are currently two modes which define how these states are stored:
+* @ref:[Distributed Data Mode](#distributed-data-mode) - uses Akka @ref:[Distributed Data](distributed-data.md) (CRDTs) (the default)
+* @ref:[Persistence Mode](#persistence-mode) - uses Akka @ref:[Persistence](persistence.md) (Event Sourcing)
```
Comment (johanandren) [inline akka-docs/src/main/paradox/cluster-sharding.md]: "which is deprecated and will be removed"?

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 6]
```diff
@@ -274,16 +276,19 @@ shard resolution, e.g. to avoid too fine grained shards. Once a shard's location
...
 
-The functionality when using the two modes is the same. If your sharded entities are not using Akka Persistence
-themselves it is more convenient to use the Distributed Data mode, since then you don't have to
-setup and operate a separate data store (e.g. Cassandra) for persistence. Aside from that, there are
-no major reasons for using one mode over the other.
+Functionally the two modes are the same. If your sharded entities are not using Akka Persistence
+themselves it is more convenient to use Distributed Data mode to not additionally
+connect a separate data store (e.g. Cassandra) for persistence.
```
Comment (johanandren) [inline akka-docs/src/main/paradox/cluster-sharding.md]: We should recommend not using the persistent one since it is going away

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 7]
```diff
@@ -323,6 +328,14 @@ akka.cluster.sharding.state-store-mode = persistence
 
 Since it is running in a cluster @ref:[Persistence](persistence.md) must be configured with a distributed journal.
 
+@@@ note
+
+Persistence mode will be replaced by a pluggable data access API with storage implementations.
+New sharding applications should no longer choose persistence mode. Existing users of persistence mode
+can eventually migrate to the replacement options.
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-docs/src/main/paradox/cluster-sharding.md]: Link to that issue in the note!

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 2 above)
Comment (patriknw) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: pro tip: add an IntelliJ code template for:
```
    /**
     * INTERNAL API
     */
    @InternalApi private[akka] 
```

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 8]
```diff
@@ -274,16 +276,19 @@ shard resolution, e.g. to avoid too fine grained shards. Once a shard's location
 the only overhead is sending a message via the `ShardRegion` rather than directly.
 
 <a id="cluster-sharding-mode"></a>
-## Distributed Data vs. Persistence Mode
+## Sharding State Store Modes
```
Comment (patriknw) [inline akka-docs/src/main/paradox/cluster-sharding.md]: I'm always somewhat on the fence if we should include documentation for deprecated features at all. We don't want new users to use it anyway, and existing users probably already know how to use it (or can look at old docs).

If it's a big feature with potentially many users I guess it makes sense to still include deprecated features in docs with clear warnings. Maybe this belongs into that category. Let's keep it in docs.

File: akka-docs/src/main/paradox/cluster-sharding.md (hunk 6 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/cluster-sharding.md]: yup, warn box

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 9]
```diff
@@ -323,6 +328,14 @@ akka.cluster.sharding.state-store-mode = persistence
 
 Since it is running in a cluster @ref:[Persistence](persistence.md) must be configured with a distributed journal.
 
+@@@ note
+
+Persistence mode will be replaced by a pluggable data access API with storage implementations.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/cluster-sharding.md]: Persistence mode for `@ref:[Remembering Entities](#remembering-entities)`

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 10]
```diff
@@ -368,6 +381,11 @@ It is always disabled if @ref:[Remembering Entities](#remembering-entities) is e
 <a id="cluster-sharding-remembering"></a>
 ## Remembering Entities
 
+Remembering entities pertains to restarting entities after a rebalance or recovering from a crash.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/cluster-sharding.md]: Good intro. Perhaps also incorporate something about that they are restarted even though no new messages are sent to them. That is the important difference. Ordinary entities are also started again, but on demand, when a new message arrives.

File: akka-docs/src/main/paradox/cluster-sharding.md [hunk 11]
```diff
@@ -401,18 +427,13 @@ you can disable durable storage and benefit from better performance by using the
...
 ```
 
-When `rememberEntities` is set to false, a `Shard` will not automatically restart any entities
-after a rebalance or recovering from a crash. Entities will only be started once the first message
+### Behavior When Not Enabled 
+
+When `rememberEntities` is disabled (the default), a `Shard` will not automatically restart any entities
+after a rebalance or recovering from a crash. Instead, entities are started once the first message
 for that entity has been received in the `Shard`. Entities will not be restarted if they stop without
 using a `Passivate`.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/cluster-sharding.md]: That last sentence is probably just adding confusion and can be removed. It will be started again when next message arrives, also if it is stopped without Passivate.

File: akka-docs/src/main/paradox/typed/persistence.md [hunk 12]
```diff
@@ -132,20 +132,6 @@ Scala
...
-resume operations @ref:[Cluster Sharding](cluster-sharding.md) is an excellent fit to spread persistent actors over a
-cluster and address them by id.
-
-The `EventSourcedBehavior` can then be run as with any plain typed actor as described in @ref:[actors documentation](actors.md),
-but since Akka Persistence is based on the single-writer principle the persistent actors are typically used together
-with Cluster Sharding. For a particular `persistenceId` only one persistent actor instance should be active at one time.
-If multiple instances were to persist events at the same time, the events would be interleaved and might not be
-interpreted correctly on replay. Cluster Sharding ensures that there is only one active entity for each id. The
-@ref:[Cluster Sharding example](cluster-sharding.md#persistence-example) illustrates this common combination.
-
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/persistence.md]: Why is this removed? This is about using Cluster Sharding together with EventSourcedBehavior, which is the canonical usage. It has nothing to do with "persistent mode" of Sharding.

Rename the heading to: Cluster Sharding and EventSourcedBehavior

**Review by patriknw — COMMENTED** [review]
looking good, a few small comments in addition to what Johan has already said

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 4 above)
Comment (helena) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: `class PersistentShardCoordinator` is not private[akka] and marked deprecated, and this private function calls that. Thus the `@silent` here.

File: akka-docs/src/main/paradox/cluster-sharding.md (hunk 8 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-sharding.md]: My thoughts, which cycled on this many, many times as I re-read the relevant docs many times, are the same. I'm only "on the fence" about it because there's this mini @raboof on my shoulder warning that until fully replaced, existing users may need it.

My gut always says we should not include docs for deprecated features. Not including them at this point helps make it clear to not use or learn to use it. So let's remove.

Comment (helena) [reply akka-docs/src/main/paradox/cluster-sharding.md]: Particularly since older docs are available and will have it.

File: akka-docs/src/main/paradox/cluster-sharding.md (hunk 5 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-sharding.md]: Persistence, if you clicked on the `#persistence-mode` link, it's noted there.

File: akka-docs/src/main/paradox/cluster-sharding.md (hunk 6 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-sharding.md]: I removed the part about setting up and operating a db  because you would almost always already have that.

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ClusterShardingSettings.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ClusterShardingSettings.scala]: Not really do-able to make this object deprecated due to how it is used and that usage breaking compile, in ways `@silent` can not fix. I think it's fine to deprecate what will be in with the second commit.

File: akka-docs/src/main/paradox/cluster-sharding.md (hunk 6 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-sharding.md]: Added warn, removed content.

File: akka-docs/src/main/paradox/typed/persistence.md (hunk 12 above)
Comment (helena) [reply akka-docs/src/main/paradox/typed/persistence.md]: Reverted/Renamed.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by raboof — APPROVED** [review]
LGTM
